src/scitacean/transfer/copy.py

```diff
@@ -0,0 +1,336 @@
+            raise FileNotAccessibleError(
+                "Cannot directly access the source folder",
+                remote_path=source_folder,
+            )
+        if not Path((source_folder / representative_file_path).posix).exists():
+            raise FileNotAccessibleError(
+                "Cannot directly access the file", remote_path=representative_file_path
+            )
+        yield CopyDownloadConnection(self._hard_link)
+
+    @contextmanager
+    def connect_for_upload(
+        self, dataset: Dataset, representative_file_path: RemotePath
+    ) -> Iterator[CopyUploadConnection]:
+        """Create a connection for uploads, use as a context manager.
+
+        Parameters
+        ----------
+        dataset:
+            The connection will be used to upload files of this dataset.
+            Used to determine the target folder.
+        representative_file_path:
+            A path on the remote to check whether files for this
+            dataset can be written.
+            The transfer assumes that, if it is possible to write to this path,
+            it is possible to write to the paths of all files to be uploaded.
+
+        Returns
+        -------
+        :
+            An open :class:`CopyUploadConnection` object.
+
+        Raises
+        ------
+        FileNotAccessibleError
+            If the remote folder cannot be accessed
+            based on ``representative_file_path``.
+        """
+        source_folder = Path(self.source_folder_for(dataset).posix)
+        if not source_folder.parents[-2].exists():
+            # This check may have a lot of false negatives.
+            # But we cannot check whether `source_folder` exists because the user
+            # may intend for the upload to create that folder.
+            # Checking the top level parent after the root should still catch many
+            # cases as long as the remote uses paths are uncommon on user machines.
+            # E.g., for /ess/data/2025/... we get parents[-2] = /ess which should
+            # not exist on non-ess machines.
+            raise FileNotAccessibleError(
+                "Cannot directly access the source folder",
+                remote_path=self.source_folder_for(dataset),
+            )
+        yield CopyUploadConnection(
+            source_folder=self.source_folder_for(dataset), hard_link=self._hard_link
+        )
+
+
+def _remote_folder_is_empty(path: RemotePath) -> bool:
+    try:
+        _ = next(iter(Path(path.posix).iterdir()))
+    except StopIteration:
+        return True
+    return False
+
+
+def _hard_link_or_copy(
+    src: os.PathLike[str] | str, dst: os.PathLike[str] | str
+) -> None:
+    try:
+        os.link(src=src, dst=dst)
+    except OSError:
+        shutil.copy(src=src, dst=dst)
+
+
+def _hard_link_or_copy2(
+    src: os.PathLike[str] | str, dst: os.PathLike[str] | str
+) -> None:
+    try:
+        os.link(src=src, dst=dst)
+    except OSError:
```

Comment:
Can we give a warning here
to inform a user that it failed to create link and made a copy instead?

Reply:
I added a log message. I don't think we need to warn here because the results will be the same except that it uses more disk space. Most people won't care.